### ISSUE TITLE
bind on eth0 instead of br-lan?

### ISSUE THREAD

**Issue by Ben-L-E** [openNDS/openNDS#49]
I'd like to run openNDS on a wired network (no wireless AP involved). I've installed openNDS v6 with libmicrohttpd-0.9.71 but when I try to bind to eth0 it fails:

```
sed -i 's/GatewayInterface br-lan/GatewayInterface eth0/' /etc/opennds/opennds.conf
systemctl start opennds
systemctl status opennds

# Aug 27 16:21:15 mysys opennds[3037]: [5][Thu Aug 27 16:21:15 2020][3038](src/util.c:242) Interface: eth0 is not yet ready - waiting...
```
(same after reboot)

This appears to be where it hangs: https://github.com/openNDS/openNDS/blob/master/src/util.c#L230-L233

My `ifname` is `eth0` and `iptype` is `inet`.  Here's my `ip address`:

```
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
    link/ether 02:07:d5:79:64:45 brd ff:ff:ff:ff:ff:ff
    inet 192.168.31.152/24 brd 192.168.31.255 scope global dynamic eth0
       valid_lft 42702sec preferred_lft 42702sec
    inet 10.20.30.40/24 brd 10.20.30.255 scope global eth0:1
       valid_lft forever preferred_lft forever
    inet6 fe80::7:d5ff:fe79:6445/64 scope link 
       valid_lft forever preferred_lft forever
```

OS: Debian Buster

Ideally I'd like to bind on `eth0:1` but that also fails.

**Comment by bluewavenet** [comment]
@Ben-L-E 
openNDS must run on a device configured as a router with at least two interfaces, running IPv4 routing with NAT between a private "lan" to a public "wan", and dhcp on the private lan to serve clients.

It is worth reading the documentation, starting with the Prerequisites section.
https://opennds.readthedocs.io/en/stable/install.html#prerequisites

**Comment by Ben-L-E** [comment]
> at least two network interfaces

I wasn't aware that excluded virtual network interfaces. I have eth0 "configured as a DHCP client, obtaining its IPv4 address and DNS server from the [ISP]" and eth0:1 configured as a "gateway in a private IPv4 subnet that is different to any private subnets between it and the ISP" as well as a DHCP and DNS server for clients with NAT configured.  That said, I understand my use case is unusual and the standard "vanilla" install requires two physical interfaces and configured with a br-lan interface so I'll close and compile with a custom fix to `util.c`.  Thanks.

**Comment by bluewavenet** [comment]
@Ben-L-E 
Virtual interfaces are not excluded. However with a single ethernet port you would have to configure two virtual switch ports with tagging for each virtual network. You would have to then physically separate them with an external managed switch. Your `ip address` output appears to show a single interface with multi-homed ip addressing.
Virtual interfaces would appear as something like eth0.1 and eth0.2

An example of a setup with virtual interfaces I have here:
```
root@OpenWrt:~# ip address
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UNKNOWN qlen 1000
    link/ether e4:95:6e:45:2e:09 brd ff:ff:ff:ff:ff:ff
    inet6 fe80::e695:6eff:fe45:2e09/64 scope link 
       valid_lft forever preferred_lft forever
4: br-lan: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP qlen 1000
    link/ether e4:95:6e:45:2e:09 brd ff:ff:ff:ff:ff:ff
    inet 10.168.1.1/24 brd 10.168.1.255 scope global br-lan
       valid_lft forever preferred_lft forever
    inet6 fdb9:a7f5:5ca4::1/60 scope global 
       valid_lft forever preferred_lft forever
    inet6 fe80::e695:6eff:fe45:2e09/64 scope link 
       valid_lft forever preferred_lft forever
5: eth0.1@eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master br-lan state UP qlen 1000
    link/ether e4:95:6e:45:2e:09 brd ff:ff:ff:ff:ff:ff
6: eth0.2@eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP qlen 1000
    link/ether e4:95:6e:45:2e:09 brd ff:ff:ff:ff:ff:ff
    inet 192.168.10.10/24 brd 192.168.10.255 scope global eth0.2
       valid_lft forever preferred_lft forever
    inet6 fe80::e695:6eff:fe45:2e09/64 scope link 
       valid_lft forever preferred_lft forever
7: wlan0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master br-lan state UP qlen 1000
    link/ether e4:95:6e:45:2e:09 brd ff:ff:ff:ff:ff:ff
    inet6 fe80::e695:6eff:fe45:2e09/64 scope link 
       valid_lft forever preferred_lft forever
```
In this case eth0.1, and wan0 are members of the bridge interface br-lan. br-lan is the interface openNDS is bound to.

eth0.2 is the wan interface.

This does show a problem when an interface is multi homed and I am pushing a fix to 6.0.1beta.
Thanks for bringing it to my attention ;-)

**Comment by Ben-L-E** [comment]
Yes, we were trying out a few configurations including IP aliasing as well as router-on-a-stick with managed switches as you noted.  Good to know the limitations and options.  Thanks again.
